File: api/app/blocks.py
"""把最短对齐切分为不可拆分差异块，并按原始 source 坐标一次性重放。

分块规则（与界面逐块批准一致）：
- 连续的非 keep 行（delete/insert）归为同一个块；keep 行天然分隔两块。
- 因此“删除与同一边界的插入”天然落在同一块内
  （Myers 平局选删除前驱，典型相邻改写为 insert 紧接 delete）。
- 块编号按对齐顺序从 0 递增；源、目标跨度以**两侧零基边界**确定，
  对同一份对齐稳定不变。

跨度采用零基半开区间 [start, end)：
- 纯删除块：source 跨度覆盖被删源项，target 跨度为空区间 [t, t)；
- 纯插入块：target 跨度覆盖新插目标项，source 跨度为空区间 [s, s)；
- 相邻改写块：两侧跨度分别覆盖被删源项与新插目标项。

重放规则（单次扫描原始对齐，绝不逐块套用导致下标漂移）：
- keep 行始终产出；
- 已批准块：insert 产出目标值，delete 不产出（采纳目标侧）；
- 未批准块：delete 产出源值，insert 不产出（保留源侧）。
于是一块不选精确等于 source，全部选中精确等于 target。
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, FrozenSet, List, Optional, Sequence, Tuple

Row = Tuple[str, Optional[int], Optional[int]]


@dataclass(frozen=True)
class Block:
    """不可拆分差异块。rows 为该块在完整对齐中的下标区间（半开）。"""

    id: int
    start_row: int
    end_row: int
    source_start: int
    source_end: int
    target_start: int
    target_end: int

    @property
    def row_range(self) -> range:
        return range(self.start_row, self.end_row)


def _row_span(
    rows: Sequence[Row], start: int, end: int, n: int, m: int
) -> Tuple[int, int, int, int]:
    """按两侧零基边界计算一块的源/目标半开跨度。

    边界来自最近的 keep 行（块前/块后），没有则取序列端点 0/N、0/M。
    """
    source_idx = [s for op, s, _ in rows[start:end] if op == "delete"]
    target_idx = [t for op, _, t in rows[start:end] if op == "insert"]

    prev_keep = next(
        (rows[i] for i in range(start - 1, -1, -1) if rows[i][0] == "keep"), None
    )
    next_keep = next(
        (rows[i] for i in range(end, len(rows)) if rows[i][0] == "keep"), None
    )

    if source_idx:
        source_start = (
            prev_keep[1] + 1 if prev_keep is not None and prev_keep[1] is not None else 0
        )
        source_end = (
            next_keep[1] if next_keep is not None and next_keep[1] is not None else n
        )
    else:
        # 纯插入：两侧边界落在同一条源边界上（空区间）。
        boundary = (
            prev_keep[1] + 1
            if prev_keep is not None and prev_keep[1] is not None
            else (next_keep[1] if next_keep is not None and next_keep[1] is not None else 0)
        )
        source_start = source_end = boundary

    if target_idx:
        target_start = (
            prev_keep[2] + 1 if prev_keep is not None and prev_keep[2] is not None else 0
        )
        target_end = (
            next_keep[2] if next_keep is not None and next_keep[2] is not None else m
        )
    else:
        # 纯删除：两侧边界落在同一条目标边界上（空区间）。
        boundary = (
            prev_keep[2] + 1
            if prev_keep is not None and prev_keep[2] is not None
            else (next_keep[2] if next_keep is not None and next_keep[2] is not None else 0)
        )
        target_start = target_end = boundary

    return source_start, source_end, target_start, target_end


def split_blocks(rows: Sequence[Row], n: int, m: int) -> List[Block]:
    """把对齐中连续的非 keep 行切分为编号稳定的差异块。"""
    blocks: List[Block] = []
    i = 0
    while i < len(rows):
        if rows[i][0] == "keep":
            i += 1
            continue
        start = i
        while i < len(rows) and rows[i][0] != "keep":
            i += 1
        end = i
        ss, se, ts, te = _row_span(rows, start, end, n, m)
        blocks.append(
            Block(
                id=len(blocks),
                start_row=start,
                end_row=end,
                source_start=ss,
                source_end=se,
                target_start=ts,
                target_end=te,
            )
        )
    return blocks


def mixed_replay(
    a: Sequence[Any],
    b: Sequence[Any],
    rows: Sequence[Row],
    accepted: FrozenSet[int] = frozenset(),
    blocks: Optional[Sequence[Block]] = None,
) -> List[Any]:
    """按块批准集合一次性重放出混合镜头序列。

    始终以原始 source 坐标 a[s]（及原始 target 坐标 b[t]）取值，
    单次扫描原始对齐：选择集合只是行级谓词，与批准先后无关，
    不会因先应用一块而漂移后续下标。

    - accepted 为空：精确得到 list(a)；
    - accepted 为全部块：精确得到 list(b)。
    """
    if blocks is None:
        blocks = split_blocks(rows, len(a), len(b))

    row_to_block: dict[int, int] = {}
    for blk in blocks:
        for r in blk.row_range:
            row_to_block[r] = blk.id

    out: List[Any] = []
    for i, (op, s, t) in enumerate(rows):
        chosen = row_to_block.get(i)
        is_accepted = chosen is not None and chosen in accepted
        if op == "keep":
            assert s is not None and t is not None and a[s] == b[t]
            out.append(a[s])
        elif op == "delete":
            assert s is not None and t is None
            if not is_accepted:
                # 未批准的删除：源镜头保留在混合序列中。
                out.append(a[s])
        elif op == "insert":
            assert t is not None and s is None
            if is_accepted:
                # 已批准的插入：目标镜头进入混合序列。
                out.append(b[t])
        else:
            raise ValueError(f"未知操作: {op}")
    return out

File: web/src/blocks.ts
import type { AlignmentRow, DiffResult } from "./types";

/**
 * 不可拆分差异块：最短对齐中连续的非 keep 行（见后端 app/blocks.py）。
 * 跨度均为零基半开区间；纯插入块源跨度为空区间 [s,s)，
 * 纯删除块目标跨度为空区间 [t,t)。块编号在同一份对齐上稳定递增。
 */
export interface DiffBlock {
  id: number;
  startRow: number;
  endRow: number;
  sourceStart: number;
  sourceEnd: number;
  targetStart: number;
  targetEnd: number;
  deletes: number[];
  inserts: number[];
}

/** 校验对齐是合法的最短对齐视图：源/目标下标零基、递增、无缺漏。 */
export function assertAlignment(result: DiffResult): void {
  let sourceIdx = 0;
  let targetIdx = 0;
  for (const row of result.alignment) {
    if (row.type === "keep") {
      if (row.source !== sourceIdx || row.target !== targetIdx) {
        throw new Error("非法的 keep 行下标");
      }
      sourceIdx += 1;
      targetIdx += 1;
    } else if (row.type === "delete") {
      if (row.source !== sourceIdx || row.target !== null) {
        throw new Error("非法的 delete 行下标");
      }
      sourceIdx += 1;
    } else if (row.type === "insert") {
      if (row.target !== targetIdx || row.source !== null) {
        throw new Error("非法的 insert 行下标");
      }
      targetIdx += 1;
    } else {
      throw new Error(`未知操作: ${(row as AlignmentRow).type}`);
    }
  }
  if (sourceIdx !== result.length_source || targetIdx !== result.length_target) {
    throw new Error("对齐未完整覆盖源或目标序列");
  }
}

interface Boundary {
  source: number;
  target: number;
}

/** 某行之前最近的 keep 边界（没有则为序列起点 0,0）。 */
function prevKeepBoundary(rows: AlignmentRow[], start: number): Boundary {
  for (let i = start - 1; i >= 0; i -= 1) {
    const row = rows[i];
    if (row.type === "keep" && row.source !== null && row.target !== null) {
      return { source: row.source + 1, target: row.target + 1 };
    }
  }
  return { source: 0, target: 0 };
}

/** 某行之后最近的 keep 边界（没有则为序列端点 N,M）。 */
function nextKeepBoundary(
  rows: AlignmentRow[],
  end: number,
  n: number,
  m: number
): Boundary {
  for (let i = end; i < rows.length; i += 1) {
    const row = rows[i];
    if (row.type === "keep" && row.source !== null && row.target !== null) {
      return { source: row.source, target: row.target };
    }
  }
  return { source: n, target: m };
}

/** 把对齐中连续的非 keep 行切分为编号稳定的差异块。 */
export function splitBlocks(result: DiffResult): DiffBlock[] {
  const { alignment: rows } = result;
  const n = result.length_source;
  const m = result.length_target;
  const blocks: DiffBlock[] = [];

  let i = 0;
  while (i < rows.length) {
    if (rows[i].type === "keep") {
      i += 1;
      continue;
    }
    const startRow = i;
    const deletes: number[] = [];
    const inserts: number[] = [];
    while (i < rows.length && rows[i].type !== "keep") {
      const row = rows[i];
      if (row.type === "delete" && row.source !== null) deletes.push(row.source);
      if (row.type === "insert" && row.target !== null) inserts.push(row.target);
      i += 1;
    }
    const endRow = i;
    const prev = prevKeepBoundary(rows, startRow);
    const next = nextKeepBoundary(rows, endRow, n, m);

    let sourceStart: number;
    let sourceEnd: number;
    if (deletes.length > 0) {
      sourceStart = prev.source;
      sourceEnd = next.source;
    } else {
      sourceStart = sourceEnd = prev.source;
    }

    let targetStart: number;
    let targetEnd: number;
    if (inserts.length > 0) {
      targetStart = prev.target;
      targetEnd = next.target;
    } else {
      targetStart = targetEnd = prev.target;
    }

    blocks.push({
      id: blocks.length,
      startRow,
      endRow,
      sourceStart,
      sourceEnd,
      targetStart,
      targetEnd,
      deletes,
      inserts,
    });
  }
  return blocks;
}

/**
 * 以原始 source 坐标一次性重放混合镜头序列。
 *
 * 选择集合只是行级谓词：单次扫描原始对齐，与勾选先后无关，
 * 不会因先应用一块而漂移后续下标。
 * - 一块不选：精确等于 source；
 * - 全部选中：精确等于 target。
 */
export function replayMixed(
  result: DiffResult,
  blocks: DiffBlock[],
  accepted: ReadonlySet<number>
): number[] {
  const { alignment: rows } = result;
  const rowToBlock = new Map<number, number>();
  for (const block of blocks) {
    for (let r = block.startRow; r < block.endRow; r += 1) {
      rowToBlock.set(r, block.id);
    }
  }

  const out: number[] = [];
  for (let i = 0; i < rows.length; i += 1) {
    const row = rows[i];
    const blockId = rowToBlock.get(i);
    const isAccepted = blockId !== undefined && accepted.has(blockId);
    if (row.type === "keep") {
      out.push(row.value);
    } else if (row.type === "delete") {
      if (!isAccepted) out.push(row.value); // 未批准删除：保留源镜头
    } else if (isAccepted) {
      out.push(row.value); // 已批准插入：采纳目标镜头
    }
  }
  return out;
}

/** 未采纳差异：为便于在界面逐块复核，把各未批准块整理成可读描述。 */
export interface PendingChange {
  blockId: number;
  kind: "delete" | "insert" | "rewrite";
  sourceStart: number;
  sourceEnd: number;
  targetStart: number;
  targetEnd: number;
}

export function pendingChanges(
  blocks: DiffBlock[],
  accepted: ReadonlySet<number>
): PendingChange[] {
  return blocks
    .filter((block) => !accepted.has(block.id))
    .map((block) => ({
      blockId: block.id,
      kind:
        block.deletes.length > 0 && block.inserts.length > 0
          ? "rewrite"
          : block.deletes.length > 0
            ? "delete"
            : "insert",
      sourceStart: block.sourceStart,
      sourceEnd: block.sourceEnd,
      targetStart: block.targetStart,
      targetEnd: block.targetEnd,
    }));
}

/** 可交付混合序列的下载文本：每行一个整数镜头编号。 */
export function mixedSequenceText(sequence: readonly number[]): string {
  return sequence.map(String).join("\n") + (sequence.length > 0 ? "\n" : "");
}
